### PR TITLE
[clang][driver] When -fveclib=ArmPL flag is in use, always link against libamath

### PR DESCRIPTION
**Pull request by pawosm-arm**
Using `-fveclib=ArmPL` without `-lamath` likely effects in the link-time errors.

### CODE REVIEW

**Comment by MacDue** [comment]
Typo in PR title `flang` -> `flag`

**Comment by pawosm-arm** [comment]
> Typo in PR title `flang` -> `flag`

Corrected both the title and the commit message. Thanks for spotting this!

File: clang/lib/Driver/ToolChains/CommonArgs.cpp
Comment (tblah) [inline clang/lib/Driver/ToolChains/CommonArgs.cpp:498]: do these flags and library names work on Windows?

**Comment by kiranchandramohan** [comment]
We probably have to provide a link to some documentation for libamath. May be something like the following one. https://developer.arm.com/documentation/102574/2410/Optimized-math-routines---libamath

File: clang/lib/Driver/ToolChains/CommonArgs.cpp
Comment (DavidTruby) [reply clang/lib/Driver/ToolChains/CommonArgs.cpp:498]: I don't _think_ that -fveclib= works on Windows in the first place. I will check though.
There should probably be an `if windows ...` check around these too regardless, in case we do enable it on future in Windows.

Comment (DavidTruby) [reply clang/lib/Driver/ToolChains/CommonArgs.cpp:498]: I suppose the actual answer to your question is no, as Windows doesn't have a separate libm, and the logic for linking to amath will probably be more complicated because you need a version of it matching your libc (the debug/release/static/dynamic grid that Windows makes you deal with)

Comment (paulwalker-arm) [reply clang/lib/Driver/ToolChains/CommonArgs.cpp:498]: Personally I think things like needing to know the dependencies required by a third party library show why we shouldn't be baking third party libraries into the driver in the first place.  How the library is built is not under llvm's control nor are the libraries part of a recognised operating system.  Are there any existing examples where other third party libraries are baked in?

Comment (DavidTruby) [reply clang/lib/Driver/ToolChains/CommonArgs.cpp:498]: We actually do support this on Windows, I was wrong above. So we should not be breaking this for Windows users.

I suppose the other option is to say that if someone passes `-fveclib=ArmPL` they will also have to pass the right flags to link to ArmPL and libamath themselves, rather than doing anything in the driver automatically? Assuming that's what @paulwalker-arm means, I'd probably agree and say that should be on the user to sort out and not on the driver.

Comment (pawosm-arm) [reply clang/lib/Driver/ToolChains/CommonArgs.cpp:498]: > do these flags and library names work on Windows?

I've reduced it to Linux only.

Comment (pawosm-arm) [reply clang/lib/Driver/ToolChains/CommonArgs.cpp:498]: > Flang and ArmPL both support Windows. I don't see why Windows should get second-class support. I would prefer to see this PR extended to also include any flags required on Windows.

This code is common for flang and clang.

@DavidTruby as an Windows expert, what flags would you recommend for Windows here?

Comment (DavidTruby) [reply clang/lib/Driver/ToolChains/CommonArgs.cpp:498]: This function you're changing here (AddLinkerInputs) isn't actually called on Windows so no need to actually add an exclusion there. ArmPL is also supported on MacOS, but there it should use the same flags as it does on Linux. So I think what you had before is fine.

You'll need something similar for supporting both clang and flang on Windows, probably somewhere around where the other Windows on Arm options are: https://github.com/llvm/llvm-project/blob/927032807dfdca5d94eb0a8707d38b605e95e407/clang/lib/Driver/ToolChains/MSVC.cpp#L86

Based on a quick glance at the ArmPL docs for the library name, the flag you'll need here is:
"--dependent-lib=amath"

Comment (paulwalker-arm) [reply clang/lib/Driver/ToolChains/CommonArgs.cpp:498]: > Flang already has a hack for a different veclib implementation on darwin

The difference with the one instance of doing this is that Accelerate is a core part of that environment, much like libm is a core part of the linux environment.  libamath is not core to any environment.  I recommend seeking advice from the wider clang community before landing this work.

Comment (pawosm-arm) [reply clang/lib/Driver/ToolChains/CommonArgs.cpp:498]: > This function you're changing here (AddLinkerInputs) isn't actually called on Windows

So much about putting 'Common' in the filename (CommonArgs.cpp)

Comment (DavidTruby) [reply clang/lib/Driver/ToolChains/CommonArgs.cpp:498]: MinGW here doesn't really mean MinGW (I realise this is quite confusing). It more specifically  means "using glibc, ld.bfd and libstdc++". This isn't supported on AArch64 at all, and is mostly being phased out on X86 too I believe. Only the "new style" mingw-w64 is supported on AArch64 (using libucrt and link.exe).

There's no Cygwin on AArch64 either.
I suppose more specifically I should have said that this function will never be called on Windows _on Arm_ 🙂

**Comment by pawosm-arm** [comment]
> We probably have to provide a link to some documentation for libamath. May be something like the following one. https://developer.arm.com/documentation/102574/2410/Optimized-math-routines---libamath

I'm always anxious to provide a link that contains numeric ID (like the one, `102574/2410`) as it suggest a dynamically generated reference that may be outlived by the lifespan of the change we're introducing in this commit...

**Comment by pawosm-arm** [comment]
I've extended it with Windows and Darwin.

**Review by tblah — APPROVED** [review]
This looks good to me but please don't merge without agreement on this approach from the other reviewers.

**Comment by pawosm-arm** [comment]
As I prepared a piece of code which makes the `-l` flags being treated properly when read from a config file (so it won't result in unresolved symbols e.g. when used with `-static`), I realized that going the config file way still creates a problem: with `-fveclib=ArmPL -lamath` put into a config file, it will always be linked with `libamath`, even if the user decides to use `-fveclib=none` to override this. And situation of this kind can only be solved properly by the patch discussed in this PR.

And there's another rationale behind it I can think of: using `-fopenmp` does not require the user to add `-lomp` to the command line, it just happens automatically. The sheer logic demands that `-fveclib=ArmPL` should result in the same treatment. There is no other provider for ArmPL's veclib functions these days than libamath.

**Comment by DavidTruby** [comment]
> with -fveclib=ArmPL -lamath put into a config file, it will always be linked with libamath, even if the user decides to use -fveclib=none

I think there might be a related issue with the approach in this patch too though, in that `-fveclib=ArmPL` will silently force linking to `amath` for _all_ math functions, rather than just the veclib functions, if we add `-lamath` as in this patch. I guess this could be resolved by changing the order of the `-l` options (i.e. have `-lm` first, so that symbols from it are selected first)?

File: clang/lib/Driver/ToolChains/CommonArgs.cpp
Comment (MaskRay) [inline clang/lib/Driver/ToolChains/CommonArgs.cpp:494]: fveclib is a Joined option that takes exactly one value. `getNumValues() == 1` is an unneeded test

File: clang/lib/Driver/ToolChains/CommonArgs.cpp
Comment (MaskRay) [inline clang/lib/Driver/ToolChains/CommonArgs.cpp:497]: We prefer no parens for `A == "xxx" && `

File: clang/test/Driver/fveclib.c
Comment (pawosm-arm) [reply clang/test/Driver/fveclib.c]: ok

File: clang/lib/Driver/ToolChains/CommonArgs.cpp
Comment (pawosm-arm) [reply clang/lib/Driver/ToolChains/CommonArgs.cpp]: ok

File: clang/lib/Driver/ToolChains/CommonArgs.cpp
Comment (pawosm-arm) [reply clang/lib/Driver/ToolChains/CommonArgs.cpp]: ok

**Comment by pawosm-arm** [comment]
I've added these suggested changes. Also, I was advised by the libamath team that the complicated relationship between libamath and libm needs more elaborate handling, hence additional `-lm` before `-lamath` (see comment that I've also added). Also I was advised to use `--as-needed`, so I wrapped it around push/pop-state.

File: clang/lib/Driver/ToolChains/CommonArgs.cpp
Comment (MaskRay) [inline clang/lib/Driver/ToolChains/CommonArgs.cpp:514]: why -m -lamath -lm ?

if there is circular dependency, the conventional way is --start-group -lamth -lm --end-group

Comment (pawosm-arm) [reply clang/lib/Driver/ToolChains/CommonArgs.cpp:514]: No no, it isn't simply a dependency. It's a matter of prioritizing system-libm over libamath which contains faster (and potentially less accurate) implementations of some of the libm functions. I was advised, these should not be used in normal circumstances (it is just unfortunate that for historical reasons, this veclib provider isn't just veclib provider; but for -fveclib=ArmPL there is no other, and likely there will be no other). From the other side, some of the libamath-only functions depend on libm functions that are not in libamath. Is it certain that start/end-group algorithm would look into libraries in-order preferring library first to provide looked upon symbol (e.g. when given `--start-group -lm -lamath --end-group`)? Besides, Darwin's default ld doesn't seem to support --start-group.

Comment (pawosm-arm) [reply clang/lib/Driver/ToolChains/CommonArgs.cpp:514]: @MaskRay is this response sufficient?

**Comment by pawosm-arm** [comment]
> > with -fveclib=ArmPL -lamath put into a config file, it will always be linked with libamath, even if the user decides to use -fveclib=none
> 
> I think there might be a related issue with the approach in this patch too though, in that `-fveclib=ArmPL` will silently force linking to `amath` for _all_ math functions, rather than just the veclib functions, if we add `-lamath` as in this patch. I guess this could be resolved by changing the order of the `-l` options (i.e. have `-lm` first, so that symbols from it are selected first)?

I missed your comment, seems it's the same issue we were discussing with libamath team, They recommended to use `-lm -lamath -lm` chain of flags.

File: clang/lib/Driver/ToolChains/CommonArgs.cpp
Comment (MaskRay) [reply clang/lib/Driver/ToolChains/CommonArgs.cpp:514]: If "It's a matter of prioritizing system-libm over libamath which contains faster (and potentially less accurate) implementations of some of the libm functions.", the comment above doesn't seem to mention this.

However, I am confused. Why cannot these duplicate symbols be removed from libamath so that the compiler driver doesn't have to play with circular -lm and -lamath dependency?

Comment (pawosm-arm) [reply clang/lib/Driver/ToolChains/CommonArgs.cpp:518]: > If "It's a matter of prioritizing system-libm over libamath which contains faster (and potentially less accurate) implementations of some of the libm functions.", the comment above doesn't seem to mention this.
> 

Now it mentions.

> However, I am confused. Why cannot these duplicate symbols be removed from libamath so that the compiler driver doesn't have to play with circular -lm and -lamath dependency?
> 
> If the reason is that we want to support both old libm and new libm where libm contains faster implementations also in libamath, I expect this reasoning to be covered.

This is purely historical reason. Libamath was meant to be libm replacement, being veclib provider for fveclib=ArmPL just one more role of this lib. The user still needs to be able to specify `-lamath` to be able to choose those less accurate faster functions, and I've made this covered by now added extra checks in the test cases.

**Review by MaskRay — APPROVED** [review]

**Pull request merged**
